### ISSUE TITLE
collectFollowSets works incorrectly when ...

### ISSUE THREAD

**Issue by xdenser** [mike-lischke/antlr4-c3#12]
I have a grammar where alternatives in one rule lead to same rule in some branches and in this case c3 fails to find alternatives.
As a synthetic  grammar  fragment example:

```
...
operand: 
           varReference
         | Number
         | functionExpr;

varRefernce: name;
funcReference: funcName LPar  arguments  RPar;
funcName: name;

name:  Id;
...

```

I think this is pretty legal grammar, but c3 will fail to find funcName rule if I want both varReference and funcName.

The problem is in collectFollowSets it cuts recursion if it finds same state on the stack (which is one to one to rule start IMHO). It is said in comment there that the algorithm is taken from ANTLR4 but there it takes into account rule context so I think there it does not cut recursion in such case.

**Comment by jasonf20** [comment]
I think I had a similar issue and disabling the use of the shortcut map helped, so that can be looked into maybe.

**Comment by xdenser** [comment]
May be reproduced with SQLite.g4 grammar and simple query text "select * from ", ask for "table_name"   rule. You will get no candidates.

**Comment by alessiostalla** [comment]
I've opened a PR for this: #54

**Comment by mike-lischke** [comment]
Fixed by PR #54.
